### DISCUSSION TITLE
Question for RequestData type

### DISCUSSION THREAD

**Discussion by CleanCoder2022** [encode/httpx#2395] [Q&A] [answered]
In `_types.py` we find this line:

`RequestData = dict`

which affects the type hint in `httpx.Request`:

```
class Request:
    def __init__(
        # ... some params omitted ...
        data: RequestData = None,
        # ... some params omitted ...
    ):
```

That means that forming a request with a string, or sending directly with one of the convenience functions, gives a type hint warning in a decent IDE:

```
already_stringified_data = json.dumps({"foo": "bar"})
client.send(Request(url, "POST", data=already_stringified_data))  # Warning: passing str instead of dict
client.post(url, data=already_stringified_data)  # Warning: passing str instead of dict
```

Looking at the implementation of Request (and PreparedRequest), it appears that using a string like this is valid usage. And it would be a pity, if you've already encoded your JSON for some other reason, to pass the original dict to httpx just so it can redundantly encode it.

If the above is a valid supported usage, I'd say that this is a bug. As the title says, RequestData ought to be Union[dict,str,bytes].

**Comment by lovelydinosaur** [comment]
That's a deprecated usage, and should be showing you a warning.

https://github.com/encode/httpx/blob/master/httpx/_content.py#L201-L206

Use `content=...` for that use case.

**Reply by blockchainGuru1018** [reply]
I agree

**Comment by blockchainGuru1018** [comment, marked as answer]
It should show a warning to you.
